### PR TITLE
administrative script

### LINKED ISSUE #274: Allow an administrator to delete pastes without digging in the file system / database
It would be nice to have an admin login option (at least one, but maybe more administrator users), and let any logged in admin delete a successfully decrypted paste. Currently, it seems you have to delete it manually from the file system or database.

### LINKED ISSUE #276: PHP-CLi tool for regularly triggering softcron independently of number of paste creations
From the configuration file, it seems like expired pastes are only removed when a new paste is created:

    [purge]
    ; minimum time limit between two purgings of expired pastes, it is only
    ; triggered when pastes are created
    ; Set this to 0 to run a purge every time a paste is created.
    limit = 300

In the installation instructions, I found no reference to having to create a cronjob or something, so I assume the above is true. That is problematic in scenarios where PrivateBin is not used often. If a new paste is not created for weeks, it can take weeks before an expired paste is removed, leaving the paste vulnerable in the meantime.

Is there a proper way to purge expired pastes automatically? Say a cronjob that calls a script or something?

### LINKED ISSUE #277: Delete empty directories in cron
Split out of https://github.com/PrivateBin/PrivateBin/issues/276

The softcron may leave empty directories on the disk.

This:
* may cause the softcron process to get ineffectively over time and potentially delete fewer pastes when it is not executed so much (see https://github.com/PrivateBin/PrivateBin/issues/276#issuecomment-361225552)
* clutters the file system
* uses useless inodes

### LINKED ISSUE #319: Simple stats ?
Hello.

As the purpose of PrivateBin is to have an anonymous mean to poaste texts, would it be possible to add stats in a simple way ?

I'm thinking of : 
- total number of pastes
- total number of expired pastes
- total number of non expired pastes
- number of expiring pastes this day/week/month...

I'm also thinking of displaying this on  homepage, with a boolean flag in config file like "display stats on home page = true/false"

In old times, some of those stats (all ?) were available with "purge.php" file which I used regularly ;) but which is unuseful now as the purge is done automatically.

Thanks :)

### PR DESCRIPTION
**Pull request by elrido**
A little script for some common or at least often requested tasks:

## Changes
* allows deleting pastes, independent of storage backend - fixes #274
* allows removing empty directories when using the Filesystem storage - fixes #277
* allows purging all expired pastes - fixes #276
* can output some statistics on all the stored pastes - fixes #319

Reg. #319 - OP suggested displaying this on the website - on a semi-large instance running those stats takes around 5 min, so it would have to get pre-generated and cached somewhere to get displayed, as it's way too slow to be updated live. Also currently the stats don't include the suggested "number of expiring pastes this day/week/month", though that would be feasible to add someone wants to.

## Further considerations

This kinda depends on #1014, because without it the purge implementation in the Filesystem will keep scanning until `PHP_INT_MAX` is reached - after that change in b8593b1bf2bb23d1946dff96450092b8bde34453 it will stop when all pastes have been enumerated and no more expired pastes are found.

Also, do we want to allow usage of this in our docker container? If yes, I'll raise a PR over there to add the php CLI to it (currently only the php FPM gets installed).

## Usage

```
Usage:
  privatebin-admin [--delete <paste id> | --empty-dirs | --help | --statistics]

Options:
  -d, --delete      deletes the requested paste ID
  -e, --empty-dirs  removes empty directories (only if Filesystem storage is
                    configured)
  -h, --help        displays this help message
  -s, --statistics  reads all stored pastes and comments and reports statistics
```

## Examples

```
# statistics of one some personal instance of mine:
$ privatebin-admin --statistics
Total:			614
................ 26% ............... 50% ................ 76% ..............
Expired:		93
Burn after reading:	12
Discussions:		31
Plain Text:		266
Source Code:		11
Markdown:		318

# statistics on our demo instance:
$ privatebin-admin --statistics
Total:			35583
................................... 10% ................................... 20% .................................. 30% ................................... 40% ................................... 50% .................................. 60% ................................... 70% ................................... 80% .................................. 90% ..................................
Expired:		933
Burn after reading:	500
Discussions:		1862
Plain Text:		30421
Source Code:		1889
Markdown:		3260

# run a purge
$ privatebin-admin --purge
purging of expired pastes concluded

# stats a few minutes after the purge
$ privatebin-admin -s
Total:			34653
................................... 10% ................................... 20% .................................. 30% ................................... 40% ................................... 50% .................................. 60% ................................... 70% ................................... 80% .................................. 90% ..................................
Expired:		8
Burn after reading:	490
Discussions:		1859
Plain Text:		29654
Source Code:		1878
Markdown:		3121
```

### CODE REVIEW

**Comment by elrido** [comment]
@rugk would you have time to review this? It's entirely new code, so I'm happy to polish this a bit more based on your feedback.

File: bin/privatebin-admin [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            if (array_key_exists('adata', $paste)) {
+                $format = $paste['adata'][1];
+                $discussion = $paste['adata'][2];
+                $burn = $paste['adata'][3];
+            } else {
+                $format = array_key_exists('formatter', $paste['meta']) ? $paste['meta']['formatter'] : 'plaintext';
+                $discussion = array_key_exists('opendiscussion', $paste['meta']) ? $paste['meta']['opendiscussion'] : false;
+                $burn = array_key_exists('burnafterreading', $paste['meta']) ? $paste['meta']['burnafterreading'] : false;
+            }
+            if ($format === 'plaintext') {
```
Comment (rugk) [inline bin/privatebin-admin:146]: That could be refactored into a private method or so to get the correct value.

**Review by rugk — APPROVED** [review]
Maybe one function per parameter/working method or one class even. Too many switches (especially for enums) are likely a sign for refactoring.

File: bin/privatebin-admin (hunk 1 above)
Comment (elrido) [reply bin/privatebin-admin]: Did extract those as recommended and wrap it into a more OOP style incl. docblocks as we do elsewhere. Also renamed it, as it is part of privatebin, so name seemed a bit redundant.

**Comment by elrido** [comment]
I noticed some inconsistencies in the usage docs of both scripts while working on https://github.com/PrivateBin/docker-nginx-fpm-alpine/pull/124 which adds support for these to the container images.

**Pull request merged**